accepted/00000-upgrading-bootstrap.md

```diff
@@ -0,0 +1,70 @@
+
+Instead of trying to manually migrate our existing very large theme source, I propose we instead try to first reduce our existing theme footprint as much as possible. We already have component scoped styles working and in use, so we can try move component or view specific styles (e.g. table styles, setup wizard etc) into component scoped styles and remove them from the global stylesheet.  
+
+This allows us to cut up the effort required and do at least some of the work gradually. I suspect the last core theme change will still be painful, though. It might be easier to rebuild the existing theme from the base Bootstrap style rather than migrating the existing one, but this is something we can consider after we've simplified the theme.  
+
+## Component and class name changes
+
+This is tied to the previous section, but worth discussing on its own. Numerous component, layout and utility classes have been dropped or changed in both upgrades, e.g. `input-group-addon` was changed in Bootstrap 4 and dropped in 5, so we need to use `input-group-text` and check if it renders correctly etc. The list is pretty long. As discussed in [this ticket](https://github.com/SUSE/spacewalk/issues/18346), it's unlikely that we'd find a tool that can handle all different ways we use Bootstrap (in JSP, JPSF, React, etc), however writing a migration script similar to the Typescript migration might be feasible. For this, we can assemble a list that covers all the breaking changes that affect us by going through the list of changes for 4 and 5, checking whether we use the affected classes, and if so, write down what the upgrade path is.  
+
+## Scripts
+
+There are some scripting changes, e.g. static method names have changed `_foo()` → `foo()`, data properties have changed `data-foo` → `data-bs-foo` etc. This mostly concerns old pages and I would expect most issues here to be straightforward to fix once we get to a state where we can run the test suite on it.
+
+## Summary
+
+In short, the suggested upgrade path can be summarized as follows:  
+
+ - Simplify the existing styles, trying to use standard Bootstrap tooling wherever possible
+ - Gather the list of breaking changes
+ - Write a migration script and/or manually migrate the above list
+ - Migrate existing styles from Less to Sass
+ - Verify the results with Cucumber
+ - Clean up any dangling ends
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Suma is huge and the UI likewise. Last time we touched the theme to match corporate styling for 4.3.0 we broke Many Things and the tail end of layout issues is long. Realistically I would expect this upgrade to require considerable time with QA and testing. In an ideal case, the upgrade would be merged as early in a new version cycle as possible, as to give everyone the maximum amount of time possible to ensure the end result is up to par. 
+
+
+# Alternatives
+[alternatives]: #alternatives
```

Comment:
How coupled we are with backend? Can we support at same time two frontends?
We might want to start developing from scratch a new frontend, without any kind of restriction not even in web frameworks, while keeping the old one as alternative.

The new frontend will not have test coverage for a while, but once we can feel all is developed, then we can switch on the test framework.

Reply:
This is a larger can of beans, but technically it would be at least somewhat possible, depending on what you mean exactly. Given the work Can has done on the HTTP API, building any kind of new UIs to match the endpoints available there would be possible. Routing-wise there wouldn't be much issue in sending certain URLs to one and and the other, we already figuratively do this with Struts and React. I think the main question to ask here would be what we would hope to get out of this idea. At some point in the past, the React based UI was this new frontend, and in that sense we are already doing it. The React parts share a lot of the baggage with the old UI such as Bootstrap and themes outlined in this RFC though.

Reply:
> [...] Can we support at same time two frontends?

FMPOV it is not really in the scope of this RFC to support two _frontend stacks_, so let's table that discussion for the time being.

A different consideration _could_ be made about supporting two _css frameworks_ - the current one and the new-and-streamlined Bootstrap 5 one. So instead of having a big-bang migration, we could ship it behind a feature flag - especially for the Uyuni community and for example our test systems - and have it more human tested before deleting the old framework completely.

Question to @Etheryte is: do you see a lot of changes necessary in the HTML structure and tag attributes, so that such an approach is not feasible/causes too much overhead?

Reply:
I think that would increase the complexity considerably, since then for every change you'd need to check both old stylesheets and new stuff that everything lines up correctly. What would be a viable alternative would be to instead have different stylesheets per page. E.g. on `/rhn/foo` you'd get the old Uyuni/Suma theme, but on `/rhn/bar` you'd get the new theme. That way we could move forward page-by-page and gradually put in the work to move things over and test them.

Reply:
I feel I don't understand what you mean, or there's some misunderstanding. I'll try to rephrase what I'm proposing and perhaps you can explain where the issue lies @moio?

What I have in mind is to have e.g. a configuration file that includes a list of paths for pages that we have updated and verified to work. For those pages, in the page header template where styles are loaded, we could load new stylesheets. For pages that are not found in that list, we can serve the old stylesheets instead.  

So for example, if we have `list-of-updated-pages.conf` such as:  

```
/rhn/foo-1
/rhn/foo-2
...
```

and then in `layout_head.jsp` a-la

```
<c:choose>
  <c:when test="${currentPage exists in list-of-updated-pages.conf}">
    <link rel="stylesheet" href="/css/new/${webTheme}.css" />
  </c:when>    
  <c:otherwise>
    <link rel="stylesheet" href="/css/old/${webTheme}.css" />
  </c:otherwise>
</c:choose>
```

This way we can create a minimal functioning stylesheet to cover one page at first, verify it works, and then that style is only served for that one page, every other page keeps working as is. Do you think this idea could work?

Reply:
That idea can work. Having half pages on the new stylesheet and the other half on the old one is, in fact, a lot better than having all pages on the new style - but with half of them broken. Anyway this is implemented, I would surely encourage to proceed page-by-page: be it with a centralized configuration file as you described or just by adjusting individual page headers to point to new stylesheets when each is ready. Let's informally call such a mechanism _developer switchable_.

The idea I was entertaining with was to make themes _user switchable_ instead, so that enterprising Uyuni community members or fearless developers could try the new theme on all pages and gauge the degree of brokenness and report bugs. What I got from your last message is that, given the coupling between Bootstrap and the markup which is expected to change from 3 to 5, in order to attain _user switchability_ we would have to maintain two copies of every page, with and without Bootstrap 5 compatibility fixes - and that does not seem like a sensible idea given the amount of code that would substantially need duplication.

Does that make sense to you, is it clearer now?

Reply:
Yes, the user switchable part was what I was missing from your idea. I think that would not be feasible because of the duplication and extra overhead. I think we're on the same page now, thanks a lot for clarifying.